### PR TITLE
fix sensor errors when device off or failed

### PR DESCRIPTION
**Pull request by nhulsch**
this is a fix for #16 based on the changes that were made by @Aohzan on the original repository (https://github.com/shenxn/ha-dyson/pull/189)

### CODE REVIEW

**Comment by dotvezz** [comment]
It's _really_ exciting to get PRs from the community after adopting it from shen, thanks for your work! This helps a lot while I've been busy with life the past couple of weeks. 

This looks good to me - I'll do some quick testing with it ASAP and if all seems good I'll get it merged and cut a new release.

File: custom_components/dyson_local/sensor.py [hunk 1]
```diff
@@ -200,10 +200,16 @@ class DysonNextDeepCleanSensor(DysonSensor):
     _attr_native_unit_of_measurement = TIME_HOURS
 
     @property
-    def native_value(self) -> int:
+    def native_value(self) -> int | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:203]:
```suggestion
    def native_value(self) -> Optional[int]
```

File: custom_components/dyson_local/sensor.py [hunk 2]
```diff
@@ -85,8 +85,8 @@ class DysonSensor(SensorEntity, DysonEntity):
     """Base class for a Dyson sensor."""
 
     _MESSAGE_TYPE = MessageType.STATE
-    _SENSOR_TYPE = None
-    _SENSOR_NAME = None
+    _SENSOR_TYPE: str | None = None
+    _SENSOR_NAME: str | None = None
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:89]:
```suggestion
    _SENSOR_TYPE: Optional[str] = None
    _SENSOR_NAME: Optional[str] = None
```

File: custom_components/dyson_local/sensor.py [hunk 3]
```diff
@@ -229,24 +242,21 @@ class DysonTemperatureSensor(DysonSensorEnvironmental):
...
     _attr_state_class = SensorStateClass.MEASUREMENT
 
-    @environmental_property
-    def temperature_kelvin(self) -> int:
-        """Return the temperature in kelvin."""
-        return self._device.temperature
-
     @property
-    def native_value(self) -> Union[str, float]:
+    def native_value(self) -> float | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:246]:
```suggestion
    def native_value(self) -> Optional[float]:
```

File: custom_components/dyson_local/sensor.py [hunk 4]
```diff
@@ -258,10 +268,17 @@ class DysonPM25Sensor(DysonSensorEnvironmental):
     _attr_native_unit_of_measurement = CONCENTRATION_MICROGRAMS_PER_CUBIC_METER
     _attr_state_class = SensorStateClass.MEASUREMENT
 
-    @environmental_property
-    def native_value(self) -> int:
+    @property
+    def native_value(self) -> float | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:272]:
```suggestion
    def native_value(self) -> Optional[float]:
```

File: custom_components/dyson_local/sensor.py [hunk 5]
```diff
@@ -273,10 +290,17 @@ class DysonPM10Sensor(DysonSensorEnvironmental):
     _attr_native_unit_of_measurement = CONCENTRATION_MICROGRAMS_PER_CUBIC_METER
     _attr_state_class = SensorStateClass.MEASUREMENT
 
-    @environmental_property
-    def native_value(self) -> int:
+    @property
+    def native_value(self) -> int | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:294]:
```suggestion
    def native_value(self) -> Optional[int]:
```

File: custom_components/dyson_local/sensor.py [hunk 6]
```diff
@@ -286,10 +310,17 @@ class DysonParticulatesSensor(DysonSensorEnvironmental):
     _attr_device_class = SensorDeviceClass.AQI
     _attr_state_class = SensorStateClass.MEASUREMENT
 
-    @environmental_property
-    def native_value(self) -> int:
+    @property
+    def native_value(self) -> int | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:314]:
```suggestion
    def native_value(self) -> Optional[int]:
```

File: custom_components/dyson_local/sensor.py [hunk 7]
```diff
@@ -300,10 +331,17 @@ class DysonVOCSensor(DysonSensorEnvironmental):
     _attr_device_class = SensorDeviceClass.AQI
     _attr_state_class = SensorStateClass.MEASUREMENT
 
-    @environmental_property
-    def native_value(self) -> float:
+    @property
+    def native_value(self) -> int | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:335]:
```suggestion
    def native_value(self) -> Optional[int]:
```

File: custom_components/dyson_local/sensor.py [hunk 8]
```diff
@@ -314,10 +352,17 @@ class DysonNO2Sensor(DysonSensorEnvironmental):
     _attr_device_class = SensorDeviceClass.AQI
     _attr_state_class = SensorStateClass.MEASUREMENT
 
-    @environmental_property
-    def native_value(self) -> int:
+    @property
+    def native_value(self) -> int | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:356]:
```suggestion
    def native_value(self) -> Optional[int]:
```

File: custom_components/dyson_local/sensor.py [hunk 9]
```diff
@@ -329,7 +374,14 @@ class DysonHCHOSensor(DysonSensorEnvironmental):
     _attr_device_class = SensorDeviceClass.VOLATILE_ORGANIC_COMPOUNDS
     _attr_state_class = SensorStateClass.MEASUREMENT
 
-    @environmental_property
-    def native_value(self) -> float:
+    @property
+    def native_value(self) -> int | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:378]:
```suggestion
    def native_value(self) -> Optional[int]:
```

File: custom_components/dyson_local/sensor.py [hunk 10]
```diff
@@ -214,10 +220,17 @@ class DysonHumiditySensor(DysonSensorEnvironmental):
     _attr_native_unit_of_measurement = PERCENTAGE
     _attr_state_class = SensorStateClass.MEASUREMENT
 
-    @environmental_property
-    def native_value(self) -> int:
+    @property
+    def native_value(self) -> int | None:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:224]:
```suggestion
    def native_value(self) -> Optional[int]:
```

File: custom_components/dyson_local/sensor.py [hunk 11]
```diff
@@ -200,10 +200,16 @@ class DysonNextDeepCleanSensor(DysonSensor):
...
         """Return the state of the sensor."""
-        return self._device.time_until_next_clean
+        if (value := self._device.time_until_next_clean) >= 0:
+            return value
+        return None
 
+    @property
+    def available(self) -> bool:
+        """Return available only if device not in off, init or failed states."""
+        return self._device.time_until_next_clean >= 0
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:212]:
```suggestion
        return isinstance(self._device.time_until_next_clean, (int, float))
```

File: custom_components/dyson_local/sensor.py [hunk 12]
```diff
@@ -214,10 +220,17 @@ class DysonHumiditySensor(DysonSensorEnvironmental):
...
         """Return the state of the sensor."""
-        return self._device.humidity
+        if (value := self._device.humidity) >= 0:
+            return value
+        return None
+
+    @property
+    def available(self) -> bool:
+        """Return available only if device not in off, init or failed states."""
+        return self._device.humidity >= 0
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:233]:
```suggestion
        return isinstance(self._device.humidity, (int, float))
```

File: custom_components/dyson_local/sensor.py [hunk 13]
```diff
@@ -229,24 +242,21 @@ class DysonTemperatureSensor(DysonSensorEnvironmental):
...
-            return temperature_kelvin
+        if (value := self._device.temperature) >= 0:
+            return value - 273.15
+        return None
 
-        return temperature_kelvin - 273.15
+    @property
+    def available(self) -> bool:
+        """Return available only if device not in off, init or failed states."""
+        return self._device.temperature >= 0
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:259]:
```suggestion
        return isinstance(self._device.temperature, (int, float))
```

File: custom_components/dyson_local/sensor.py [hunk 14]
```diff
@@ -273,10 +290,17 @@ class DysonPM10Sensor(DysonSensorEnvironmental):
...
         """Return the state of the sensor."""
-        return self._device.particulate_matter_10
+        if (value := self._device.particulate_matter_10) >= 0:
+            return value
+        return None
+
+    @property
+    def available(self) -> bool:
+        """Return available only if device not in off, init or failed states."""
+        return self._device.particulate_matter_10 >= 0
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:303]:
```suggestion
        return isinstance(self._device.particulate_matter_10, (int, float))
```

File: custom_components/dyson_local/sensor.py [hunk 15]
```diff
@@ -286,10 +310,17 @@ class DysonParticulatesSensor(DysonSensorEnvironmental):
...
         """Return the state of the sensor."""
-        return self._device.particulates
+        if (value := self._device.particulates) >= 0:
+            return value
+        return None
+
+    @property
+    def available(self) -> bool:
+        """Return available only if device not in off, init or failed states."""
+        return self._device.particulates >= 0
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:323]:
```suggestion
        return isinstance(self._device.particulates, (int, float))
```

File: custom_components/dyson_local/sensor.py [hunk 16]
```diff
@@ -300,10 +331,17 @@ class DysonVOCSensor(DysonSensorEnvironmental):
...
         """Return the state of the sensor."""
-        return self._device.volatile_organic_compounds
+        if (value := self._device.volatile_organic_compounds) >= 0:
+            return value
+        return None
+
+    @property
+    def available(self) -> bool:
+        """Return available only if device not in off, init or failed states."""
+        return self._device.volatile_organic_compounds >= 0
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:344]:
```suggestion
        return isinstance(self._device.volatile_organic_compounds, (int, float))
```

File: custom_components/dyson_local/sensor.py [hunk 17]
```diff
@@ -314,10 +352,17 @@ class DysonNO2Sensor(DysonSensorEnvironmental):
...
         """Return the state of the sensor."""
-        return self._device.nitrogen_dioxide
+        if (value := self._device.nitrogen_dioxide) >= 0:
+            return value
+        return None
+
+    @property
+    def available(self) -> bool:
+        """Return available only if device not in off, init or failed states."""
+        return self._device.nitrogen_dioxide >= 0
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:365]:
```suggestion
        return isinstance(self._device.nitrogen_dioxide, (int, float))
```

File: custom_components/dyson_local/sensor.py [hunk 18]
```diff
@@ -329,7 +374,14 @@ class DysonHCHOSensor(DysonSensorEnvironmental):
...
+    @property
+    def native_value(self) -> int | None:
         """Return the state of the sensor."""
-        return self._device.formaldehyde
+        if (value := self._device.formaldehyde) >= 0:
+            return value
+        return None
+
+    @property
+    def available(self) -> bool:
```
Comment (dotvezz) [inline custom_components/dyson_local/sensor.py:385]:
```suggestion
        return isinstance(self._device.time_until_next_clean, (int, float))
```

**Review by dotvezz — CHANGES_REQUESTED** [review]
Whew, sorry for the slow review here. And of course thanks for your contribution! I've just got two categories of suggestions here. One is admittedly a nitpick, but the other is likely a bit more important.

1. For all the `int | None` type hints, let's use `Optional[int]` to be idiomatic and consistent with the rest of the (admittedly slow-growing bunch of) hints.
2. In my testing environment, I ran into TypeErrors on the `available` properties. Let's just be a bit more explicit on those checks.

Once that's taken care of, we're good to go!

**Comment by nhulsch** [comment]
Feel free to change it directly in your repository. I was just adopting code from a PR on the original repo

**Comment by dotvezz** [comment]
@nhulsch thanks again for this! I finally got around to updating and releasing it a few days ago. Welcome to the contributors club!

image

**Pull request closed without merging**